[AArch64] Initial sched model for Neoverse N3

FLZ101 opened the pull request:
References:

* Arm Neoverse N3 Software Optimization Guide
* Arm A64 Instruction Set for A-profile architecture

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

References:

* Arm Neoverse N3 Software Optimization Guide
* Arm A64 Instruction Set for A-profile architecture

FLZ101 commented on llvm/include/llvm/TargetParser/ARMTargetParser.def:
```diff
@@ -380,6 +380,9 @@ ARM_CPU_NAME("neoverse-n1", ARMV8_2A, FK_CRYPTO_NEON_FP_ARMV8, false,
 ARM_CPU_NAME("neoverse-n2", ARMV9A, FK_NEON_FP_ARMV8, false,
              (ARM::AEK_BF16 | ARM::AEK_DOTPROD | ARM::AEK_FP16FML |
               ARM::AEK_I8MM | ARM::AEK_RAS | ARM::AEK_SB ))
+ARM_CPU_NAME("neoverse-n3", ARMV9_2A, FK_NEON_FP_ARMV8, false,
```
Thanks. The changes are undone.

FLZ101 commented:
@c-rhodes @sjoerdmeijer @rj-jesus

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+                                         "^UMOVvi(8|16|32|64)$")>;
+
+// ASIMD transfer, gen reg to element
+def : InstRW<[N3Write_5c_1M0_2V], (instregex "^INSvi(8|16|32|64)gpr$")>;
```
Why 2V? Same for other places.

FLZ101 replied:
From "Neoverse N3 Software Optimization Guide":

| Instruction Group                  | AArch64 Instructions | Exec Latency | Execution Throughput | Utilized Pipelines |
| ---------------------------------- | -------------------- | ------------ | -------------------- | ------------------ |
| ASIMD transfer, gen reg to element | INS                  | 5            | 1                    | M0, V              |

* V has two ports V0, V1 and so can execute two uops per cycle. 
* "Execution Throughput" is 1, so only one instruction can be completed per cycle. And there are at most two uops for V decoded from that instruction.
* Here "2V" means two uops for V.

davemgreen replied:
I think the bottleneck on throughput comes from elsewhere, maybe the M0 pipeline? The optimization guide does not contain all the details, but I believe the core should be able, for example, to execute an INS and another V operation in the same cycle.

FLZ101 replied:
Actually I used a script to generate a draft of the sched model based on a simple rule: matching the throughput assuming all utilized ports are fully utilized without considering any microarchitecture details.

Are those details described in some documents ?

FLZ101 replied:
Done. With V2 as a reference, those '2V' has been replaced with '1V'.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+
+// Arithmetic, shift by immediate
+// Arithmetic, shift by immediate and insert
+// def : InstRW<[N3Write_2c_1V1], (instrs SHRNB, SHRNT, SSHLLB, SSHLLT, USHLLB, USHLLT)>;
```
Commented code.

FLZ101 replied:
done.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+def : InstRW<[N3Write_8c_4V_6L], (instregex "^LD4D_IMM$")>;
+
+// Contiguous Load four structures to four vectors, scalar + imm
+def : InstRW<[N3Write_12c_5V_7L], (instregex "^LD4[BHW]_IMM$")>;
```
Where do these various (5V, 7L) numbers come from?

FLZ101 replied:
Here "5V" means 5 uops for V and "7L" means 7 uops for L. How I get these is described in the comment above explaining how I get "2V".

davemgreen replied:
So kind of random guessing? Would 4L + 8V work OK too?

FLZ101 replied:
(mV, nL) where m <= 5, n <= 7 should work.

I choose those numbers to match the throughput assuming all utilized ports are fully utilized.

davemgreen replied:
There may be some level of approximation in the SWOG values. If there are already a breakdown in the V2 scheduling model then that would probably be a good guide to follow: https://github.com/llvm/llvm-project/blob/6023d17e6b6624913b85fe9d2b5d79ae681e5764/llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td#L2673

FLZ101 replied:
In the V2 SWOG:

Instruction Group | AArch64 Instructions | Exec Latency | Execution Throughput | Utilized Pipelines
-- | -- | -- | -- | --
Contiguous Load four structures to four vectors, scalar + imm  | LD4B, ... | 9 | 1/2 | V, L

To match the throughput, the write should be `V2Write_9cyc_mL_nV` where `(m == 6, n <= 8)` or `(m <= 6, n == 8)`. To determine m and n, we need to either know more microarchitecture details or conduct some experiments in a real machine.

FLZ101 replied:
That is also the case for N3. Since I have neither access to those microarchitecture details nor a real N3 machine, there is nothing more I can do to improve the write here.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+
+// Multiply accumulate, W-form
+// Multiply accumulate long
+def : SchedAlias<WriteIM32, N3Write_2c_1M0>;
```
A mul (a madd without a addend) seem to be able to execute on either M pipeline.

FLZ101 replied:
It seems that there are some errors and omissions in "Neoverse N3 Software Optimization Guide".

**From "Neoverse N2 Software Optimization Guide"**:

![{85EEA1F2-D9C3-4BA2-94BE-4F7E0B7B4EDE}](https://github.com/user-attachments/assets/d4d70a73-b98d-4571-8d2c-bd6b4f8686c4)

**From "Neoverse N2 Software Optimization Guide"**:

![{83236E78-EB30-4A76-9FFE-88FAD1ADEF3F}](https://github.com/user-attachments/assets/3285dd7d-0ce9-4a1a-a7f1-e4749b52e804)

FLZ101 replied:
I need to compare "Neoverse N3 Software Optimization Guide" with "Neoverse N2 Software Optimization Guide" further to fix all similar errors.

davemgreen replied:
In the N3 manual it is a bit hidden in Note 3 below the table that says:
```
3. Multiply without accumulate when Ra is ZR (0'b11111), MUL, MNEG, SMULL, SMNEGL, UMULL and UMNEGL
instructions can be executed on utilized pipeline M with an execution throughput of 2.
```

FLZ101 replied:
Sure.

FLZ101 replied:
Done.

davemgreen commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -1127,7 +1127,7 @@ def : ProcessorModel<"neoverse-n1", NeoverseN1Model,
                      ProcessorFeatures.NeoverseN2, [TuneNeoverseN2]>;
 def : ProcessorAlias<"cobalt-100", "neoverse-n2">;
-def : ProcessorModel<"neoverse-n3", NeoverseN2Model,
+def : ProcessorModel<"neoverse-n3", NeoverseN3Model,
```
Can you change the Cortex-A725 to use the same scheduling model too. I think this will be a closer match than the NeoverseN2Model.

FLZ101 replied:
done.

davemgreen commented on llvm/unittests/TargetParser/Host.cpp:
```diff
@@ -85,6 +85,9 @@ TEST(getLinuxHostCPUName, AArch64) {
   EXPECT_EQ(sys::detail::getHostCPUNameForARM("CPU implementer : 0x41\n"
                                               "CPU part        : 0xd0c"),
             "neoverse-n1");
+  EXPECT_EQ(sys::detail::getHostCPUNameForARM("CPU implementer : 0x41\n"
```
You can commit this missing test separately.

FLZ101 replied:
done. I've created a separate PR https://github.com/llvm/llvm-project/pull/106749

davemgreen reviewed (commented):
I didn't look at everything but most of what I looked at matches what I would have expected. I just had some questions/comments.

FLZ101 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+def : InstRW<[N3Write_2c_1L01_1V], (instregex "^STUR[BHSD]i$")>;
+
+// Store vector reg, unscaled immed, Q-form
+def : InstRW<[N3Write_2c_1L01_1V], (instrs STURQi)>;
```
done.

FLZ101 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+def : ReadAdvance<ReadVLD,     0>;
+
+def : WriteRes<WriteAtomic,  []> { let Unsupported = 1; }
+def : WriteRes<WriteFDiv,    []> { let Unsupported = 1; }
```
Since they are not used by the sched model. Is it better to set them even they are not used?

davemgreen replied:
I think I would copy what other models do. If there are new instructions added with these Write's (which granted might be unlikely for some), then the scheduling info would get some sensible default for the new instructions, even if the model doesn't know about them. (Although I can see your point about them being unused).

FLZ101 replied:
Done.

davemgreen commented:
Hello. The other thing that came up from talking to internal colleagues was whether we have any evidence that this scheduling model is working well and reflecting the real hardware. Have you run any performance data or even have you looked into calibrating against the latencies/throughputs produced from hardware?

FLZ101 commented:
> Hello. The other thing that came up from talking to internal colleagues was whether we have any evidence that this scheduling model is working well and reflecting the real hardware. Have you run any performance data or even have you looked into calibrating against the latencies/throughputs produced from hardware?

I have no N3 machine. I just "translate" the N3 SWOG into a sched model.

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

FLZ101 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2359 @@
+def N3Write_1c_2I : SchedWriteRes<[N3UnitI]> {
+  let Latency = 1;
+  let NumMicroOps = 2;
+  let ReleaseAtCycles = [2];
```
Done.

davemgreen commented:
> I have no N3 machine. I just "translate" the N3 SWOG into a sched model.

OK. I will see if I can try and run something to give it a quick check.

One extra thing, this definition says "take one N3UnitL and hold it for 6 cycles and one N3UnitV is held for 4 cycles".
```
def N3Write_8c_6L_4V : SchedWriteRes<[N3UnitL, N3UnitV]> {
    let Latency = 8;
    let NumMicroOps = 10;
    let ReleaseAtCycles = [6, 4];
}
```

If the operation is split down into micro-operations then they should be separate uops that can be executed independantly. So this version for example says "there are 4 L ops and 8 v ops that can execute in any unit as needed".
```
def V2Write_9cyc_4L_8V : SchedWriteRes<[V2UnitL, V2UnitL, V2UnitL,
                                        V2UnitL, V2UnitV, V2UnitV,
                                        V2UnitV, V2UnitV, V2UnitV,
                                        V2UnitV, V2UnitV, V2UnitV]> {
  let Latency     = 9;
  let NumMicroOps = 12;
}
```

The dependency between the L and V ops is not well described in tablegen scheduling models, but it should lead to better analysis in general. Things like divide (and in older models sqrt) generally use ReleaseAtCycles, others operations should ideally split them out.

FLZ101 commented:
> OK. I will see if I can try and run something to give it a quick check.

Thanks. I will have more time to work on this MR this weekend.

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

* use ReleaseAtCycles only for "Divide, [WX]-form"

FLZ101 commented:
> Things like divide (and in older models sqrt) generally use ReleaseAtCycles, others operations should ideally split them out.

Thanks. Now I only keep `ReleaseAtCycles` for "Divide, [WX]-form".

In the N3 SWOG, only  integer divides are performed using an iterative algorithm and operations like "FP divide and square root", "ASIMD FP divide and square root", etc  are now performed using a fully pipelined data path.

FLZ101 commented:
> Things like divide (and in older models sqrt) generally use ReleaseAtCycles, others operations should ideally split them out.

Thanks. Now I only keep `ReleaseAtCycles` for "Divide, [WX]-form".

In the N3 SWOG, only  integer divides are performed using an iterative algorithm and operations like "FP divide and square root", "ASIMD FP divide and square root", etc  are now performed using a fully pipelined data path.

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

* Multiply without accumulate when Ra is ZR, MUL, MNEG, SMULL, SMNEGL, UMULL and UMNEGL
instructions can be executed on utilized pipeline M with an execution throughput of 2.

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+                                         "^LDNT1S[BH]_ZZR_S$")>;
+
+// Non temporal gather load, vector + scalar 64-bit element size
+def : InstRW<[N3Write_6c_3L], (instregex "^LDNT1S?[BHW]_ZZR_D$")>;
```
I would expect 2L here.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+def : InstRW<[N3Write_8c_1V_1L], (instregex "^LD2[BHWD]$")>;
+
+// Contiguous Load three structures to three vectors, scalar + imm
+def : InstRW<[N3Write_8c_3V_4L], (instregex "^LD3D_IMM$")>;
```
3L

FLZ101 replied:
done.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+def : InstRW<[N3Write_8c_3V_4L], (instregex "^LD3D_IMM$")>;
+
+// Contiguous Load three structures to three vectors, scalar + imm
+def : InstRW<[N3Write_10c_6V_9L], (instregex "^LD3[BHW]_IMM$")>;
```
9L sounds like a lot. I would expect 3L most of the time.

FLZ101 replied:
done.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+def : InstRW<[N3Write_11c_6V_9L_12I], (instregex "^LD3[BHW]$")>;
+
+// Contiguous Load four structures to four vectors, scalar + imm
+def : InstRW<[N3Write_8c_4V_6L], (instregex "^LD4D_IMM$")>;
```
4L for the LD4's.

FLZ101 replied:
done.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+                                         "^GLD(FF)?1W_IMM$")>;
+
+// Gather load, vector + imm, 64-bit element size
+def : InstRW<[N3Write_6c_3L], (instregex "^GLD(FF)?1S?[BHW]_D_IMM$",
```
I would expect a 2-element gather (v2i64) to use 2 L ops, but maybe some extra work is needed to combine them back together.

davemgreen reviewed (approved):
> Thanks. Now I only keep ReleaseAtCycles for "Divide, [WX]-form".
> 
> In the N3 SWOG, only integer divides are performed using an iterative algorithm and operations like "FP divide and square root", "ASIMD FP divide and square root", etc are now performed using a fully pipelined data path.

Yeah that is nice. Thanks for the updates, I managed to run some testing and it looked OK. It took a little while to set up but the performance, overall, was a very slight improvement (but that might be in the realms of noise). It helps validate that nothing is going too wrong with the new numbers.

I have some suggestions mostly about the number of L pipes used in certain operations. I don't think I would worry too much about hitting the throughput numbers in the SWOG exactly so long as they are close. There can be other things other than number of micro-ops that can cause bottlenecks that are not modelled as well in the scheduler. It probably makes more sense to try and keep the micro-ops sensible.

If you can make these adjustments then this LGTM. Thanks for working on this and making the changes.

FLZ101 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+def : InstRW<[WriteAdr, N3Write_6c_2L], (instregex "LD2Rv(16b|8h|4s|2d)_POST$")>;
+
+// ASIMD load, 3 element, multiple, D-form, B/H/S
+def : InstRW<[N3Write_8c_4L_3V], (instregex "LD3Threev(8b|4h|2s)$")>;
```
done.

N3Write_8c_3L_3V still matches the throughput.

FLZ101 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2375 @@
+def : InstRW<[WriteAdr, N3Write_6c_3L], (instregex "LD3Rv(16b|8h|4s|2d)_POST$")>;
+
+// ASIMD load, 4 element, multiple, D-form, B/H/S
+def : InstRW<[N3Write_8c_6L_4V], (instregex "LD4Fourv(8b|4h|2s)$")>;
```
done.

FLZ101 committed:
[AArch64] Initial sched model for Neoverse N3

* adjust some Ls and Is

FLZ101 commented:
> I am checking others with N2ScheModel and V2SchedModel as references. May take some time.

Done. Besides changing some `L`s, I also adjust some `I`s and perform a rebase.

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -0,0 +1,2333 @@
+    let NumMicroOps = 9;
+}
+
+def N3Write_4c_3L01_3I_3V : SchedWriteRes<[N3UnitL01, N3UnitL01, N3UnitL01,
```
I think they should all be 1I, as in there is 1 additional integer operation to calculate the address.

FLZ101 replied:
I am not certain which one is better.

AArch64SchedNeoverseN2.td:

```
// Contiguous store three structures from three vectors, scalar + scalar
def : InstRW<[N2Write_7c_5L01_5S_5V], (instrs ST3H)>;
```

AArch64SchedNeoverseV2.td:

```
// Contiguous store three structures from three vectors, scalar + scalar
def : InstRW<[V2Write_7c_9L01_9S_9V01], (instregex "^ST3[BHWD]$")>;
```

It seems that the number of integer uops is the same as the number of vector uops. So I set the number of `I` to 3.

davemgreen replied:
Oh I see. They might be wrong too, but I'm just guessing from what seems plausible and we don't have any official documentation one way or the other. Lets keep it like this then for the moment, and if we need to change it we can change them all.

Are you happy for me to hit submit?

FLZ101 replied:
Sure. Thanks.

davemgreen reviewed (approved):
If you can make that last change (or explain why it isn't needed), then this LGTM and we can get it in. Thanks
